modules/031-linstor/images/linstor-pools-importer/main.go

```diff
@@ -0,0 +1,386 @@
+	if err != nil {
+		klog.Fatalln("Failed to create event", err)
+	}
+}
+
+// Makes loop over storage pool candidates, retruns channel of changed ones
+func runCandidatesLoop(ctx context.Context, f func() ([]lclient.StoragePool, error), delay time.Duration) <-chan lclient.StoragePool {
+	ch := make(chan lclient.StoragePool)
+	go func() {
+		var oldCandidates []lclient.StoragePool
+		for {
+			select {
+			case <-ctx.Done():
+				close(ch)
+				return
+			default:
+				candidates, err := f()
+				if err != nil {
+					klog.Fatalln("failed to load candidates:", err)
+				}
+			LOOP:
+				for _, candidate := range candidates {
+					for _, oldCandidate := range oldCandidates {
+						if candidate.StoragePoolName == oldCandidate.StoragePoolName {
+							continue LOOP
+						}
+					}
+					ch <- candidate
+				}
+				oldCandidates = candidates
+			}
+			time.Sleep(delay)
+		}
+	}()
+	return ch
+}
+
+type VolumeGroups struct{}
+
+func getLVMThinStoragePools() ([]lclient.StoragePool, error) {
+	cmd := exec.Command("vgs", "-oname,tags", "--separator=;", "--noheadings", "--config="+lvmConfig)
+
+	var out bytes.Buffer
+	cmd.Stdout = &out
+	err := cmd.Run()
+	if err != nil {
+		return nil, err
+	}
+	return parseVolumeGroups(out.String())
+}
+
+type ThinPools struct{}
+
+func getLVMStoragePools() ([]lclient.StoragePool, error) {
+	cmd := exec.Command("lvs", "-oname,vg_name,lv_attr,tags", "--separator=;", "--noheadings", "--config="+lvmConfig)
+	var out bytes.Buffer
+	cmd.Stdout = &out
+	err := cmd.Run()
+	if err != nil {
+		return nil, err
+	}
+	return parseThinPools(out.String())
+}
+
+// Collects all storage pool candidates from the node
+func getCandidates() ([]lclient.StoragePool, error) {
+	var candidates []lclient.StoragePool
+
+	// Getting LVM storage pools
+	c, err := getLVMStoragePools()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get LVM storage pools: %s", err)
+	}
+	for _, vg := range c {
+		candidates = append(candidates, vg)
+	}
+
+	// Getting LVM thin storage pools
+	c, err = getLVMThinStoragePools()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get LVMthin storage pools: %s", err)
+	}
+	for _, tp := range c {
+		candidates = append(candidates, tp)
+	}
+
+	return candidates, nil
+}
+
+func parseThinPools(out string) ([]lclient.StoragePool, error) {
+	var sps []lclient.StoragePool
+	for _, line := range strings.Split(out, "\n") {
+		if line == "" {
+			continue
+		}
+		a := strings.Split(line, ";")
+		if len(a) != 4 {
+			return nil, fmt.Errorf("wrong line: %s", line)
+		}
+		lvName, vgName, lvAttr, tags := strings.TrimSpace(a[0]), a[1], a[2], strings.Split(a[3], ",")
+		if lvName == "" {
+			return nil, fmt.Errorf("LV name can't be empty")
+		}
+		if vgName == "" {
+			return nil, fmt.Errorf("vgName can't be empty")
+		}
+		if lvAttr == "" {
+			return nil, fmt.Errorf("lvAttr can't be empty")
+		}
+		if lvAttr[0:1] != "t" {
+			continue
+		}
+		name := parseNameFromLVMTags(&tags)
+		if name == "" {
+			continue
+		}
```

Comment:
> We should at least return an error here

This is regular situation. Node can have other LVM groups and thinpools not intended for LINSTOR use.
Eg the system ones for root filesystem and so on. These pools have no any tag assigned, so they can simple be skipped.

Actually, `parseThinPools()` just parses `lvs` output and returns ready-to-use objects (`lclient.StoragePool`) to the upstream loop.

There is `runCandidatesLoop()` which is continuously running `getCandidates()` to read storage pools configuration from the node, it also performs a check with previous output to not overwhelm API with excessive requests.

Another words:
- If we implement any logging at this level it will repeat the same message each 10 seconds, which is not good.
- If we would move all checks and logging to upper level we would need to transfer significantly more objects and handle them each time on the upper level.

Reply:
> If we implement any logging at this level it will repeat the same message each 10 seconds, which is not good.

Ok. What is worse: ignore erroneous desired configuration or report junk? I won't choose 😆

> If we would move all checks and logging to upper level we would need to transfer significantly more objects and handle them each time on the upper level.

Even if you mean API calls, I would not be bothered by it right now. Optimizations would be better done for a real-world reason, then from assumptions.

To sum up, I'd rather avoid of loosing time on implicitness, and prefer grepping out explicit log junks. I'd stick to more explicit data flow with errors. But I won't insist anyway.

Reply:
> Ok. What is worse: ignore erroneous desired configuration or report junk? I won't choose 😆

I think there is not much options to make erroneous desired configuration.

We have just simple check: if pool has `linstor-somename` tag assigned
- If not then it is just skipped
- If it has then it must be configured.

The only problem I can see, is tag name collisions: eg. when we detect two pools with same tag assigned or when we detect single pool with two tags assigned. I'm going to add check for this on upper level.